# Finding scope 3 emissions of a Pharma company.
- Scope 3 category is downstream transportation and distribution
- Data taken from data.world
- Distance-based methology will be used for calculations
    - According to GHG "If sub-contractor fuel data cannot be easily obtained in order to use the fuel-based method, then the distance-based method should be used."

# Data points required for calculations of scope 3
1. Total distance in KM 
    - Coordinates of places
2. Size of packages
3. Mode of Transport
4. Emission Factors
5. Calculation formulaes
    1. Emissions from road transport: = ∑ (mass of goods purchased (tonnes) × distance travelled in transport leg × emission factor of transport mode or vehicle type (kg CO2e/tonne-km)
    2. emissions from air transport: = ∑ (quantity of goods purchased (tonnes) x distance travelled in transport leg x emission factor of transport mode or vehicle type (kg CO2e/tonne-km))
    3. emissions from sea transport: = ∑ (quantity of goods purchased (tonnes) x distance travelled in transport leg x emission factor of transport mode or vehicle type (kg CO2e/tonne-km))

### Steps I took
1. Searching for dataset
2. Understanding the dataset and making calculative assumptions
    - Assumptions made should have proper justification.
3. Listing down the data points required for calculations
4. Listing down the calculation formulaes
5. Filtering out the data columns required from main data
6. Exploratory Data Analysis
7. Choosing a single year for calculations


## Data columns
1. Country - Destination location
2. Mode - Mode of transport (Air, Ocean, Road)
3. Weight - Kilogram of weight
4. Delivery Date - Date of delivery
5. Manufacturing Site - Source location

In [108]:
# Import libraries
import pandas as pd
import numpy as np
import geopandas as gpd

# For geolocation
from geopy.geocoders import ArcGIS

# For calculation of source airport/port
from math import radians, sin, cos, sqrt, atan2

# Calculating the time
import time

# For converting country name to iso_2 codes
import country_converter as coco

In [109]:
# Collecting data
raw_df = pd.read_csv("./Data/SCMS_Delivery_History_Dataset_20150929.csv")

In [111]:
# Filtering out required columns from the main data.

filtered_df = raw_df[['ID','Country','Shipment Mode','Manufacturing Site','Weight (Kilograms)','Delivery Recorded Date','Item Description']].copy()
filtered_df.rename(columns={"Shipment Mode":"Mode","Weight (Kilograms)":"Weight","Delivery Recorded Date":"Delivery_Date",'Manufacturing Site':'Manufacturing_site','Country':'Destination_Country'},inplace=True)

In [112]:
# Converting Delivery Date to datetime datatype
filtered_df.loc[:,'Delivery_Date'] = pd.to_datetime(filtered_df['Delivery_Date'])

# Exploratory Data Analysis

In [113]:
# Checking total number of data points for each year
filtered_df['Delivery_Date'].groupby(by=filtered_df['Delivery_Date'].dt.year).size()

Delivery_Date
2006      65
2007     670
2008    1109
2009    1192
2010    1176
2011    1049
2012    1250
2013    1205
2014    1599
2015    1009
Name: Delivery_Date, dtype: int64

In [114]:
# Check for null values
filtered_df.isna().sum()

ID                       0
Destination_Country      0
Mode                   360
Manufacturing_site       0
Weight                   0
Delivery_Date            0
Item Description         0
dtype: int64

Null values found in the mode of transport. 

In [115]:
# Checking Mode column's null values

# filtered_df[filtered_df['Mode'].isna()]

# Counting the number of null values per year
filtered_df[filtered_df['Mode'].isna()].groupby(by=filtered_df['Delivery_Date'].dt.year).size()

Delivery_Date
2006      2
2007    264
2008     94
dtype: int64

Since mode of transport is not available for these years (2006,2007,2008) I will not consider these years for our base year.
Even though year 2006 has only 2 NaN rows for Mode, the number of data points are only 65. Thus not enough to consider as base year as compared to other years.

In [116]:
# Checking Weight column values
filtered_df[filtered_df['Weight'].str.isnumeric() == False]

ID Destination_Country   Mode             Manufacturing_site  \
8         46             Nigeria    Air      Aurobindo Unit III, India   
12        62             Nigeria    Air           EY Laboratories, USA   
15        68            Zimbabwe    Air             BMS Meymac, France   
16        69             Nigeria    NaN  ABBVIE GmbH & Co.KG Wiesbaden   
31       262        South Africa    NaN       GSK Mississauga (Canada)   
...      ...                 ...    ...                            ...   
10318  86817            Zimbabwe  Truck              Cipla, Goa, India   
10319  86818            Zimbabwe  Truck     Mylan,  H-12 & H-13, India   
10320  86819       C�te d'Ivoire  Truck   Hetero Unit III Hyderabad IN   
10321  86821              Zambia  Truck  Cipla Ltd A-42 MIDC Mahar. IN   
10323  86823            Zimbabwe  Truck              Cipla, Goa, India   

                           Weight Delivery_Date  \
8           See ASN-93 (ID#:1281)    2006-12-07   
12     Weight Captured Separately    2007-01-10   
15     Weight Captured Separately    2007-03-19   
16     Weight Captured Separately    2007-05-07   
31     Weight Captured Separately    2008-01-29   
...                           ...           ...   
10318     See DN-4307 (ID#:83920)    2015-07-20   
10319     See DN-4307 (ID#:83920)    2015-07-20   
10320     See DN-4313 (ID#:83921)    2015-08-07   
10321  Weight Captured Separately    2015-09-03   
10323  Weight Captured Separately    2015-08-11   

                                        Item Description  
8                      Stavudine 30mg, capsules, 60 Caps  
12           HIV 1/2, InstantChek HIV 1+2 Kit, 100 Tests  
15     #102198**Didanosine 200mg [Videx], tablets, 60...  
16        HIV 1/2, Determine Complete HIV Kit, 100 Tests  
31     Zidovudine 10mg/ml [Retrovir], oral solution, ...  
...                                                  ...  
10318  Lamivudine/Nevirapine/Zidovudine 30/50/60mg, d...  
10319  Lamivudine/Nevirapine/Zidovudine 30/50/60mg, d...  
10320  Lamivudine/Zidovudine 150/300mg, tablets, 60 Tabs  
10321  Efavirenz/Lamivudine/Tenofovir Disoproxil Fuma...  
10323  Lamivudine/Zidovudine 30/60mg, dispersible tab...  

[3952 rows x 7 columns]

As you can see above the weight column has non-numeric values 'Weight Captured Separately', thus the values for these rows are not available. While values like 'See ASN-93 (ID#:1281)' can be found by mapping to that particular ID and getting the weight

In [117]:
# Counting the number of 'Weight Captured Separately' values in weight columns per year.

filtered_df[(filtered_df['Weight'].str.isnumeric() == False) & (filtered_df['Weight'] == 'Weight Captured Separately')].groupby(by=filtered_df['Delivery_Date'].dt.year).size()

Delivery_Date
2006      6
2007     28
2008    233
2009    262
2010     67
2011     56
2012     33
2013     80
2014    386
2015    356
dtype: int64

Seeing that 2012 has least number of 'Weight Captured Separately' values and it has good number of data points i.e. 1250. We can consider it as our base year.

# Data Cleaning

## Fixing Weight column

In [118]:
# Filtering out 'Weight Captured Separately' rows from Final dataset
final_df = filtered_df[(filtered_df['Delivery_Date'].dt.year == 2012) & (filtered_df['Weight'] != 'Weight Captured Separately')].copy()
final_df.reset_index(inplace = True, drop = True)

In [119]:
# Function to resolve "See DN-2947 (ID#:83642)" given in weight column and adding the weight

def mapWeights(weight):
    try:
        if weight.isnumeric() == False:
            ID = weight[-6:-1]
            weight_returned = filtered_df[filtered_df['ID'] == int(ID)]['Weight'].iloc[0]
            if weight_returned == 'Weight Captured Separately':
                return None
#             print(f'{ID} -- {weight_returned}')
            return float(weight_returned)
        return float(weight)
    except Exception as e:
        print(f'Error == {e} \n {weight[-6:-1]} --- {filtered_df[filtered_df["ID"] == int(weight[-6:-1])]["Weight"].iloc[0]}')

In [120]:
weights = final_df['Weight'].apply(mapWeights)

In [121]:
# Adding weights list to the df
final_df.loc[:,'Weight'] = weights

In [122]:
# Removing None value rows for weight

final_df = final_df[final_df['Weight'].isna() == False]

In [123]:
# Converting Weights from KG to Tonnes
# 1 kilogram = 0.001 tonne

final_df['Weight'] = final_df['Weight'] * 0.001

## Fixing Mode column

In [124]:
final_df['Mode'].unique()

array(['Air', 'Ocean', 'Truck', 'Air Charter'], dtype=object)

As you can see we have air charter as the 4th Mode of transport. We can combine that with Air transport

In [125]:
final_df.loc[final_df['Mode'] == 'Air Charter','Mode'] = 'Air'

## Getting Location coordinates of Source Location and Destination Location

### If Mode of Transport is,
1. Air and Truck, then we get the Airport coordinates of the Capital City of the Country given
2. Ocean, then we get the coordinates of the main port of the country.

## To-Do
- Source 
1. Get the appropriate location name, port/airport.
2. Get the coordinates
- Destination
1. Get the coordinates

In [126]:
# Getting airports data
raw_airport_df = pd.read_csv('Data/airports.csv')
# Read ports data
raw_ports_df = gpd.read_file('Data/attributed_ports.geojson')

In [127]:
# Getting required columns for 
df = raw_airport_df[['type','latitude_deg','longitude_deg','name','iso_country','municipality']]
final_airport_df = df[df['type'].str.lower().str.contains('airport')]

final_ports_df = raw_ports_df[['Country','Name','geometry']].copy()
final_ports_df['latitude_deg'] = final_ports_df.geometry.apply(lambda p: p.y)
final_ports_df['longitude_deg'] = final_ports_df.geometry.apply(lambda p: p.x)
final_ports_df.drop(['geometry'], axis=1,inplace=True)
final_ports_df['iso_country'] = coco.convert(final_ports_df['Country'],to='ISO2', not_found=None)

In [128]:
final_airport_df

type  latitude_deg  longitude_deg                       name  \
1       small_airport     38.704022    -101.473911       Aero B Ranch Airport   
2       small_airport     59.947733    -151.692524               Lowell Field   
3       small_airport     34.864799     -86.770302               Epps Airpark   
5       small_airport     34.942803     -97.818019             Fulton Airport   
6       small_airport     34.305599    -112.165001             Cordes Airport   
...               ...           ...            ...                        ...   
74638  medium_airport     40.542524     122.358600      Yingkou Lanqi Airport   
74639  medium_airport     41.784401     123.496002    Shenyang Dongta Airport   
74641   small_airport    -11.584278      47.296389  Glorioso Islands Airstrip   
74642   small_airport     32.110587     -97.356312      Fainting Goat Airport   
74644   small_airport     30.784722     130.270556     Satsuma Iōjima Airport   

      iso_country      municipality  
1              US             Leoti  
2              US      Anchor Point  
3              US           Harvest  
5              US              Alex  
6              US            Cordes  
...           ...               ...  
74638          CN  Laobian, Yingkou  
74639          CN  Dadong, Shenyang  
74641          TF  Grande Glorieuse  
74642          US              Blum  
74644          JP           Mishima  

[44441 rows x 6 columns]

In [153]:
# Initialising geolocation object
nom = ArcGIS()
# DF to append Source location
data = {'Manufacturing_site':[],'latitude':[],'longitude':[],'Mode':[]}
source_locations = pd.DataFrame(data)

In [130]:
# Functions to get the distance in KM
def distance(lat1, lon1, lat2, lon2):
    # Calculate the distance between two coordinates using the haversine formula
    """
    Calculate the great circle distance between two points
    on the earth (specified in decimal degrees)

    All args must be of equal length.    

    """
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = np.sin(dlat/2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2.0)**2

    c = 2 * np.arcsin(np.sqrt(a))
    km = 6367 * c
    return km

In [154]:
"""
Function to check 
Check in source location DF if the place and mode are already calculated.
If not calculated, it means that for this place and mode the source has not been calculated.
If already calculated, then skip to next manufacturing site
"""
def checkSourceLocation(place,mode):
    if source_locations.size == 0:
        return True
    for index,row in source_locations.iterrows():
           if row['Manufacturing_site'] == place and row['Mode'] == mode:
                   return False
    return True

In [155]:
"""
Function to find the minimum distance
"""
def minDistance(transport_df,location,iso2_country):
    min_distance = float('inf')
    min_latitude = float('inf')
    min_longitude = float('inf')

    # Getting all airports lying inside the country
    transport_df_country = transport_df[transport_df['iso_country'] == iso2_country]

    for index,row in transport_df_country.iterrows():
        dist = distance(location.latitude, location.longitude, row['latitude_deg'],row['longitude_deg'])
        if dist <= min_distance:
            min_distance = dist
            min_latitude = row['latitude_deg']
            min_longitude = row['longitude_deg']

    return min_latitude,min_longitude

In [156]:
# Function to get the coordinates of the location
# Now that we have location of the manufacturing site coordinates. We need to get the Airport/Port location from which the item was transported
def getSourceLocation(data):
    # Code to Measure time taken by program to execute.
    # store starting time
    begin = time.time()
#   print(data)
    place = data['Manufacturing_site']
    mode = data['Mode']
    try:
        
        # Check if source location has already been calculated or not
        if checkSourceLocation(place,mode):
            location = nom.geocode(place)
            country = nom.reverse(str(location.latitude) + ',' + str(location.longitude)).address.split(',')[-1].replace(" ", "")
            # Getting iso_2 code for the country
            iso2_country = coco.convert(names=country, to='ISO2',not_found=None)
      
            # Now that we have location of the manufacturing site. We need to get the Airport/Port location from which the item was transported
            if mode == 'Air':
                # Finding the airport which has minimum distance
                min_latitude,min_longitude = minDistance(final_airport_df,location,iso2_country)
                source_locations.loc[len(source_locations.index)] = [place,min_latitude,min_longitude,mode]
            elif mode == 'Ocean':
                # Finding the port which has minimum distance
                min_latitude,min_longitude = minDistance(final_ports_df,location,iso2_country)
                source_locations.loc[len(source_locations.index)] = [place,min_latitude,min_longitude,mode]
            else:
                source_locations.loc[len(source_locations.index)] = [place,location.latitude,location.longitude,mode]
        else:
            print(f'Already Completed {place} {mode}')
    except Exception as e:
        print(e, place)
        source_locations.loc[len(source_locations.index)] = [place,None,None,mode]

In [157]:
# Adding source_latitude and source_longitude columns
# store starting time
begin = time.time()
final_df.apply(getSourceLocation,axis=1)
end = time.time()
# total time taken
print(f"Total runtime of the program is {end - begin}")

Already Completed Alere Medical Co., Ltd. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Trinity Biotech, Plc Air
Already Completed Trinity Biotech, Plc Air
Already Completed Chembio Diagnostics Sys. Inc. Air


2643 not found in ISOnumeric


Already Completed Aurobindo Unit III, India Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Trinity Biotech, Plc Air
Already Completed Aurobindo Unit III, India Air
Already Completed Aurobindo Unit III, India Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Chembio Diagnostics Sys. Inc. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Trinity Biotech, Plc Air
Already Completed Hetero Unit III Hyderabad IN Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Aurobindo Unit III, India Air
Already Completed KHB Test Kit Facility, Shanghai China Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Cipla, Goa, India Air
Already Completed Inverness Japan Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Chembio Diagnostics Sys. Inc. Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Alere Medical Co., Ltd

충청북도옥천군청산면교평리 not found in regex


Already Completed Aurobindo Unit III, India Air
Already Completed Aspen-OSD, Port Elizabeth, SA Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Chembio Diagnostics Sys. Inc. Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Trinity Biotech, Plc Air
Already Completed ABBVIE (Abbott) Logis. UK Air
Already Completed MSD, Haarlem, NL Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Mylan (formerly Matrix) Nashik Air
Already Completed Aurobindo Unit III, India Air
Already Completed Alere Medical Co., Ltd. Air
Already Completed MSD, Haarlem, NL Air
Already Completed Hetero Unit III Hyderabad IN Air
Already Completed MSD, Haarlem, NL Air
Already Completed Cipla, Goa, India Air
Already Completed Trinity Biotech, Plc Air
Already Completed Alere Medical Co., Ltd. Air
Already Complete

The code took around 74 seconds to complete the computations

Merging the source_location_df to the final_df

In [237]:
final_df = final_df.merge(source_locations, how='outer', on=['Manufacturing_site','Mode'])

In [238]:
final_df

ID Destination_Country Mode          Manufacturing_site  Weight  \
0     12973               Haiti  Air           Cipla, Goa, India   0.021   
1     19105              Rwanda  Air           Cipla, Goa, India   1.631   
2     39441              Zambia  Air           Cipla, Goa, India   5.999   
3     46005              Rwanda  Air           Cipla, Goa, India   6.490   
4     46574              Rwanda  Air           Cipla, Goa, India   0.049   
...     ...                 ...  ...                         ...     ...   
1197  84152         South Sudan  Air  Macleods Daman Plant INDIA   2.652   
1198  84166               Libya  Air  Macleods Daman Plant INDIA   0.709   
1199  85257  Dominican Republic  Air  Macleods Daman Plant INDIA   0.544   
1200  85866          Congo, DRC  Air  Macleods Daman Plant INDIA   0.895   
1201  86463              Uganda  Air  Macleods Daman Plant INDIA   1.421   

     Delivery_Date                                   Item Description  \
0       2012-06-12  Lamivudine/Nevirapine/Stavudine 30/50/6mg, dis...   
1       2012-06-22    Abacavir 20mg/ml, oral solution, Bottle, 240 ml   
2       2012-06-26  Lamivudine/Nevirapine/Stavudine 60/100/12mg, d...   
3       2012-06-28  Lamivudine 10mg/ml, oral solution, Bottle, 240 ml   
4       2012-06-15  Lamivudine/Stavudine 60/12mg, dispersible tabl...   
...            ...                                                ...   
1197    2012-03-08                 Lamivudine 150mg, tablets, 60 Tabs   
1198    2012-05-23                 Lamivudine 150mg, tablets, 60 Tabs   
1199    2012-05-04                 Lamivudine 150mg, tablets, 60 Tabs   
1200    2012-09-11                 Lamivudine 150mg, tablets, 60 Tabs   
1201    2012-03-27                 Lamivudine 150mg, tablets, 60 Tabs   

       latitude  longitude  
0     15.380800  73.831398  
1     15.380800  73.831398  
2     15.380800  73.831398  
3     15.380800  73.831398  
4     15.380800  73.831398  
...         ...        ...  
1197  20.434401  72.843201  
1198  20.434401  72.843201  
1199  20.434401  72.843201  
1200  20.434401  72.843201  
1201  20.434401  72.843201  

[1202 rows x 9 columns]

Checking for any NONE or INF values in longitude and latitude columns

In [239]:
final_df.loc[(final_df['latitude'].isna()) | (final_df['latitude'] == float('inf'))]

ID Destination_Country Mode           Manufacturing_site  Weight  \
669  16085               Haiti  Air     Medochemie Factory A, CY   1.173   
745  24246            Ethiopia  Air    ABBVIE (Abbott) Logis. UK   1.050   
746  26019              Rwanda  Air    ABBVIE (Abbott) Logis. UK   0.253   
747  29677             Vietnam  Air    ABBVIE (Abbott) Logis. UK   0.275   
748  37960            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.374   
749  37979            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.374   
750  44577            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.014   
751  51217            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.254   
752  65987             Vietnam  Air    ABBVIE (Abbott) Logis. UK   0.164   
753  68268            Ethiopia  Air    ABBVIE (Abbott) Logis. UK   0.728   
754  77750            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.309   
755  81358            Zimbabwe  Air    ABBVIE (Abbott) Logis. UK   1.374   
843  75595               Haiti  Air  Standard Diagnostics, Korea   0.002   

    Delivery_Date                                   Item Description  \
669    2012-01-18  Mefloquine 250mg (as hydrochloride), tablets, ...   
745    2012-06-27  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
746    2012-06-29  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
747    2012-11-29  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
748    2012-12-13  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
749    2012-11-08  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
750    2012-12-13  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
751    2012-04-02  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
752    2012-06-27  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
753    2012-02-20  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
754    2012-11-22  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
755    2012-11-08  Lopinavir/Ritonavir 80/20mg/ml [Kaletra], oral...   
843    2012-01-23          Malaria Antigen P.f , HRP2, Kit, 25 Tests   

     latitude  longitude  
669       inf        inf  
745       NaN        NaN  
746       NaN        NaN  
747       NaN        NaN  
748       NaN        NaN  
749       NaN        NaN  
750       NaN        NaN  
751       NaN        NaN  
752       NaN        NaN  
753       NaN        NaN  
754       NaN        NaN  
755       NaN        NaN  
843       inf        inf

In [240]:
indexes = final_df.loc[(final_df['latitude'].isna()) | (final_df['latitude'] == float('inf'))].index

Since coordinates of these manufacturing sites were not found, I will drop these.

In [241]:
final_df.drop(index=indexes,axis=1,inplace=True)

In [242]:
final_df.loc[(final_df['latitude'].isna()) | (final_df['latitude'] == float('inf'))]

Empty DataFrame
Columns: [ID, Destination_Country, Mode, Manufacturing_site, Weight, Delivery_Date, Item Description, latitude, longitude]
Index: []

Adding suffixes to latitude and longitude columns

In [243]:
final_df.rename(columns={'latitude':'source_lat','longitude':'source_log','Manufacturing_site':'Source_site'},inplace=True)

In [244]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1189 entries, 0 to 1201
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID                   1189 non-null   int64         
 1   Destination_Country  1189 non-null   object        
 2   Mode                 1189 non-null   object        
 3   Source_site          1189 non-null   object        
 4   Weight               1189 non-null   float64       
 5   Delivery_Date        1189 non-null   datetime64[ns]
 6   Item Description     1189 non-null   object        
 7   source_lat           1189 non-null   float64       
 8   source_log           1189 non-null   float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 92.9+ KB
